instrgen/docs/design.md

```diff
@@ -0,0 +1,200 @@
+# OpenTelemetry Go Source Automatic Instrumentation - design
+
+This document describes design of `instrgen` - compile time automatic OpenTelemetry instrumentation for Go programming language.
+
+## Background
+
+Usually techniques used for instrumenting programs in specific languages are connected to runtimes and type systems
+used by these languages and depend on capabilities offered by them. Methods differs depending on language properties.
+Are they executed by virtual machine? Are they statically or dynamically typed?
+Language and its internals influence instrumentation techniques.
+In dynamic languages it's very easy to replace original implementation. This is done by providing a new
+implementation built on top of the original one. A technique called monkey patching (i.e. Python decorators).
+
+Go is statically typed, natively compiled programming language which means that works on bare methal without
+virtual machine in between. The only component that lives between your program and machine is operating system.
+Because of this limitation, auto-instrumentation of Go programs is limited to the following:
+
+1. Auto-instrument within the kernel itself (i.e. [eBPF](https://ebpf.io/))
+2. Auto-instrument during compilation of the program
+
+The second approach is taken by `instrgen`.
+
+## Rationale
+
+Compile time instrumentation is not different by other mechanisms described above.
+The fundamental difference is the injection point, instrumentation has to be done 
+at compile time which has cons and pros.
+
+Go is perfect language for compile time instrumentation for two main reasons:
+- everything that's needed is in standard library
+- it's easy to control compilation process due to go compiler architecture
+
+## Design
+
+As mentioned above `instrgen` utilizes toolexec compiler switch (https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4058/).
+
+**NOTE** `driver` executable needs to be accessible via $PATH
+
+### Flow
+`instrgen` driver has two passes. First pass prepares some configuration for second pass based on provided settings,
+it means that the same process is executed twice in different context.
+
+It was designed and implemented this way to take input directly from user via command line switches and to do some
+initial analysis, however this can be spit into two separate tools or processes.
+
+`driver --inject  /testdata/basic yes main.main`
+
+Executable takes few parameters:
+
+- command name
+- file pattern - determines files taken into instrumentation
+- yes/no - determines whether code should be injected in-place
+- entry point - a function from which injection process start
+
+```sh
+usage driver --command [file pattern] replace entrypoint
+    command:
+	inject                                 (injects open telemetry calls into project code)
+	prune                                  (prune open telemetry calls
```

Comment:
what does it mean to prune opentelemetry calls?  Which parts of opentelemetry will it remove (I assume it removes things)?

Reply:
For now, `instrgen` works in two modes, it can instrument user code in place or instrument intermediate code generated during normal go compiler compilation phase (this changes are invisible to the user). Prune relates to the first one.